### ISSUE TITLE
Make RowIterator countable

### ISSUE THREAD

**Issue by dkarlovi** [openspout/openspout#99]
A common thing to do when reading Excel files is finding out how much there is to read. Making the RowIterator countable can do that in an efficient way instead of making downstream do it.

**Comment by Slamdunk** [comment]
Well, the nature of this library is to be efficient, and to be so is to read the next row only when the previous row has already been processed.

How would you know the number of rows if you can't read them all before consumption?

**Comment by dkarlovi** [comment]
I'm currently debugging the profile of this library with a file which makes reading it extremely slow. One cause is the data processing is done upfront, and not delegated (ie row is fully parsed just to skip to the next row).

If the row iterator were countable, it could know if the data processing is needed or not.

**Comment by Slamdunk** [comment]
I repeat, how would you know the number of rows before reading them?

**Comment by dkarlovi** [comment]
You still read the rows, you just don't need to fully process them. For example, creating the shared strings cache just to get a number.

**Comment by Slamdunk** [comment]
Any PR is welcome :+1:

**Comment by dkarlovi** [comment]
Yes, I'm also looking at why the library is using 500MB to read a 12MB file, it might make sense to first figure that one out, then we can see how this one makes sense.

Will create another issue.

**Comment by dkarlovi** [comment]
@Slamdunk maybe a better alternative is to make `\OpenSpout\Reader\XLSX\RowIterator::getCell` calling `\OpenSpout\Reader\XLSX\Helper\CellValueFormatter::extractAndFormatNodeValue` lazy?

I still think a higher level API which shows intent is better (because the lib can optimize for it), but this could be some sort of half-step.

**Comment by Slamdunk** [comment]
> I'm also looking at why the library is using 500MB to read a 12MB file

This library is optimized to avoid such issues, checkout https://github.com/openspout/openspout/tree/4.x/benchmarks folder.
Openspout reads 300.000 Excel rows with less than 5MB of RAM

**Comment by dkarlovi** [comment]
Maybe some Excel files, but this one was created by Excel itself, not an Excel writer library, note I made it read only 500 rows too, the files contains thousands:

![Untitled-profile-Blackfire](https://user-images.githubusercontent.com/209225/194083634-8fc7e04d-b9ce-4808-abca-2b5ada373402.png)

**Comment by Slamdunk** [comment]
The number of calls has nothing to do, as far as I can tell, with the occupied RAM.

Please add your Excel file and your script to the benchmarks to prove this library is the culprit.

**Comment by dkarlovi** [comment]
I cannot since this file is full of proprietary information.

The library is 100% the culprit, so far there's several issues:
1. I'm fetching a count of rows (here 10576), it only needs to run through the rows and and count them, but it will create the shared strings cache (expensive) and process & create each cell of each row, as noted in the above comment, this would be solved with making the RowIterator countable and doing it in an efficient way internally
2. it seems the library doesn't handle having a lot of shared strings and string being long (kilobytes), in my file there's 13k of them, the cache looks like this:
```
/app $ ls -lh /tmp/sharedstrings633d93e63b192/
total 65M    
-rw-rw-rw-    1 2000     2000        6.2M Oct  5 14:25 sharedstrings0
-rw-rw-rw-    1 2000     2000        6.8M Oct  5 14:25 sharedstrings1
-rw-rw-rw-    1 2000     2000      294.7K Oct  5 14:25 sharedstrings10
-rw-rw-rw-    1 2000     2000      286.0K Oct  5 14:25 sharedstrings11
-rw-rw-rw-    1 2000     2000        6.0M Oct  5 14:25 sharedstrings12
-rw-rw-rw-    1 2000     2000        6.3M Oct  5 14:25 sharedstrings13
-rw-rw-rw-    1 2000     2000        5.9M Oct  5 14:25 sharedstrings2
-rw-rw-rw-    1 2000     2000        5.9M Oct  5 14:25 sharedstrings3
-rw-rw-rw-    1 2000     2000        5.9M Oct  5 14:25 sharedstrings4
-rw-rw-rw-    1 2000     2000        5.6M Oct  5 14:25 sharedstrings5
-rw-rw-rw-    1 2000     2000        5.6M Oct  5 14:25 sharedstrings6
-rw-rw-rw-    1 2000     2000        5.2M Oct  5 14:25 sharedstrings7
-rw-rw-rw-    1 2000     2000        4.7M Oct  5 14:25 sharedstrings8
-rw-rw-rw-    1 2000     2000      307.1K Oct  5 14:25 sharedstrings9
```

**Comment by simstern** [comment]
I was looking at this as I do have the same issue. I tried a few things and came up with this:
https://github.com/swisscomeventandmedia/openspout/commit/8c1b56f98b71ee0afa5da3ee54cd0f1cffa6406d

I'm not completely sold on this solution as it changes the iterator pointer. On the other hand, I do not have a better idea without digging too deep.

Any feedback is welcome. If this is a feasible solution, I'm happy to do something similar for the CSV and ODS readers and before submitting a PR.
